docs/architecture/adr-055-orm.md

```diff
@@ -0,0 +1,105 @@
+# ADR 055: ORM
+
+## Changelog
+
+* 2022-04-27: First draft
+
+## Status
+
+ACCEPTED Implemented
+
+## Abstract
+
+In order to make it easier for developers to build state machines and for clients to query, index and verify proofs
+against state data, we have implemented an ORM (object-relational mapping) layer for the Cosmos SDK.
+
+## Context
+
+Historically modules in the Cosmos SDK have always used the key-value store directly and created various handwritten
+functions for managing key format as well as constructing secondary indexes. This consumes a significant amount of
+time when building a module and is error-prone. Because key formats are non-standard, sometimes poorly documented,
+and subject to change, it is hard for clients to generically index, query and verify merkle proofs against state data.
+
+The known first instance of an "ORM" in the Cosmos ecosystem was in [weave](https://github.com/iov-one/weave/tree/master/orm).
+A later version was built for [regen-ledger](https://github.com/regen-network/regen-ledger/tree/master/orm) for
+use in the group module and later [ported to the SDK](https://github.com/cosmos/cosmos-sdk/tree/main/x/group/internal/orm)
+just for that purpose.
+
+While these earlier designs made it significantly easier to write state machines, they still required a lot of manual
+configuration, didn't expose state format directly to clients, and were limited in their support of different types
+of index keys, composite keys, and range queries.
+
+Discussions about the design continued in https://github.com/cosmos/cosmos-sdk/discussions/9156 and more
+sophisticated proofs of concept were created in https://github.com/allinbits/cosmos-sdk-poc/tree/master/runtime/orm
+and https://github.com/cosmos/cosmos-sdk/pull/10454.
+
+## Decision
+
+These prior efforts culminated in the creation of the Cosmos SDK `orm` go module which uses protobuf annotations
+for specifying ORM table definitions. This ORM is based on the new `google.golang.org/protobuf/reflect/protoreflect`
+API and supports:
+* sorted indexes for all simple protobuf types (except `bytes`, `enum`, `float`, `double`) as well as `Timestamp` and `Duration`
```

Comment:
why we don't support sorted bytes index?

Reply:
The existing sorting over strings covers all use cases I'd have in mind

Reply:
If we store address as bytes then it will be useful.

Reply:
But what does sorting get us with addresses? They're not intrinsically ordered